### PR TITLE
Use AirflowJsonEncoder in BaseXCom.serialize_value

### LINKED ISSUE #16386: SageMaker operators return non-serializable value
**Apache Airflow version**: 1.10.15, 2.1.0

**What happened**:
The `execute` methods of all SageMaker operators return a dictionary which contains `datetime` objects (e.g. [here](https://github.com/apache/airflow/blob/69a1a732a034406967e82a59be6d3c019e94a07b/airflow/providers/amazon/aws/operators/sagemaker_training.py#L120) but affects all operators). Execution fails consistently, throwing `TypeError: Object of type 'datetime' is not JSON serializable`.

There is already a ticket for this issue [here](https://issues.apache.org/jira/browse/AIRFLOW-3772), I am opening this to add visibility in GitHub and reference it in an eventual PR.

**What you expected to happen**:
The operator's `execute` should return only serializable values. I suggest that these operators should return only the name of the job, instead of the full description, which can then be used elsewhere to pull further information about the job if needed.

### PR DESCRIPTION
**Pull request by olivermeyer**
closes: #16386

Use the custom `AirflowJsonEncoder` when serializing a value for XCom.

### CODE REVIEW

**Review by potiuk — APPROVED** [review]
LGTM but I would love to have someone more familiar with the encoder to verify if there are no side effects ? Things like performance and some potential serialize/deserialize compatibility come to mind immmediately.

**Comment by olivermeyer** [comment]
> LGTM but I would love to have someone more familiar with the encoder to verify if there are no side effects ? Things like performance and some potential serialize/deserialize compatibility come to mind immmediately.

Thanks for reviewing. Do you know who is familiar with the encoder and could take a look?

Also, CI says `The PR likely needs to run all tests!` - is that the case? If yes, I'll amend and force-push. If not, is there a way to trigger CI without force-pushing? One of the tests failed but it passes locally so could be a transient error.

**Comment by potiuk** [comment]
> Thanks for reviewing. Do you know who is familiar with the encoder and could take a look?

Let's see who will chime in :). There are a few CODEOWNERS already added automatically so I guess one or few of those nice people will.

> Also, CI says `The PR likely needs to run all tests!` - is that the case? If yes, I'll amend and force-push. If not, is there a way to trigger CI without force-pushing? One of the tests failed but it passes locally so could be a transient error.

Please rebase and push. This is pretty standard and you will likely expect to do couple of those while you iterate on your PRs in Airflow.

**Comment by olivermeyer** [comment]
Bumping this PR - any chance one of the reviewers could take a look?

**Comment by potiuk** [comment]
Hey Everyone - any comments here? @kaxil :) ? I somehow have a feeling that using the custom encoder is not the best idea.

**Review by kaxil — COMMENTED** [review]
Yeah please hold off on this one, as a part of https://github.com/apache/airflow/issues/8703 we were hoping/planning to use Serialization code in https://github.com/apache/airflow/blob/19ed074e9c696eb4aff25f3a833e7f359f2b1c38/airflow/serialization/serialized_objects.py to achieve the same and more.

And at some point remove the JsonEncode in airflow.utils in favor of it too.

**Comment by olivermeyer** [comment]
@kaxil thanks for taking a look. I'll close this PR for now. One thing isn't clear to me: will there be a need for a similar PR in the future to use some other custom serialization when setting an Xcom, or will that somehow be handled as part of #8703?

**Comment by kaxil** [comment]
> @kaxil thanks for taking a look. I'll close this PR for now. One thing isn't clear to me: will there be a need for a similar PR in the future to use some other custom serialization when setting an Xcom, or will that somehow be handled as part of #8703?

Yup, that is correct, there was an attempt in https://github.com/apache/airflow/pull/9847 however the PR was abandoned

**Pull request closed without merging**
